modules/dnn/src/int8layers/quantization_utils.cpp

```diff
@@ -35,43 +35,122 @@ static void broadcast1D2TargetMat(Mat& data, const MatShape& targetShape, int ax
+    CV_Assert(axis >= 0 && axis < src.dims());
+    CV_Assert(repetitions > 1);
+
+    Mat src_mat = src.getMat();
+    Mat dst_mat;
+
+    MatShape sshape = shape(src_mat);
+    MatShape dshape = sshape;
+
+    size_t dtype_bytes = src_mat.elemSize();
+    int chunk_size = dtype_bytes;
+    int num_chunks = 1;
+
+    dshape[axis] *= repetitions;
+
+    for (int i = axis+1; i < src.dims(); ++i)
+        chunk_size*=sshape[i];
+
+    for (int i = 0; i <= axis; ++i)
+        num_chunks*=sshape[i];
+
+    dst.create(src.dims(), dshape.data(), src.type());
+    dst_mat = dst.getMat();
 
-    zeropointsMat.create(subTargetShape.size(), subTargetShape.data(), CV_32FC1);
-    scalesMat.create(subTargetShape.size(), subTargetShape.data(), CV_32FC1);
 
-    const int len = scales.size();
-    // Deep copy the scales and zeropoint data and prevent the original data from being changed.
+    for (int i = 0; i < repetitions; ++i) {
+        size_t src_offset = 0;
+        size_t dst_offset = i * chunk_size;
+
+        for (int j = 0; j < num_chunks; ++j) {
+            memcpy(dst_mat.data + dst_offset, src_mat.data + src_offset, chunk_size);
+            src_offset += chunk_size;
+            dst_offset += chunk_size * repetitions;
+        }
+    }
+}
+
+template <typename T>
+static void copyVecToMat(Mat& mat, const std::vector<T>& data){
+    float * matPtr = mat.ptr<float>(0);
+    const int len = data.size();
 
-    float * scalePtr = scalesMat.ptr<float>(0);
     for (int i = 0; i < len; i++)
-        scalePtr[i] = scales[i];
+        matPtr[i] = (float) data[i];
+}
 
-    float * zpPtr = zeropointsMat.ptr<float>(0);
+template <typename T>
+static void copyMatToVec(const Mat& mat, std::vector<T>& data){
+    const float * matPtr = mat.ptr<float>(0);
+    const int len = mat.total();
+    data.clear();
     for (int i = 0; i < len; i++)
-        zpPtr[i] = (float )zeropoints[i];
+        data.push_back(matPtr[i]);
+}
+
+template <typename T>
+static void broadcastBlockedMatrix(Mat& mat, const std::vector<T>& data, const MatShape& targetShape, int axis, int block_size){
+    CV_Assert(targetShape[axis] % block_size == 0 && block_size <= targetShape[axis]);
+
+    MatShape subTargetShape(targetShape);
+    subTargetShape[axis] = static_cast<int>(subTargetShape[axis] / block_size);
+    Mat tmpMat(subTargetShape.size(), subTargetShape.data(), CV_32FC1);
 
-    broadcast1D2TargetMat(scalesMat, targetShape, axis);
-    broadcast1D2TargetMat(zeropointsMat, targetShape, axis);
+    copyVecToMat(tmpMat,data);
+
+    block_repeat(tmpMat, axis, block_size, mat);
+}
+
+template <typename T>
+static void broadcastStandardMatrix(Mat& mat, const std::vector<T>& data, const MatShape& targetShape, int axis)
+{
+    MatShape subTargetShape(targetShape.size(), 1);
+    subTargetShape[axis] = data.size();
+    mat.create(subTargetShape.size(), subTargetShape.data(), CV_32FC1);
+
+    copyVecToMat(mat,data);
+
+    broadcast1D2TargetMat(mat, targetShape, axis);
+}
```

Comment:
I propose to drop `template` as we are operating a `CV_32FC1` Mat either way.

---

If you want to keep template, you will need to distinguish between types and use `char` along with elementwise to do the correct copy.

Reply:
I decided to use `template` because `std::vector<T>& data` can be either `vector<float>` in the case of scale or `vector<int>` in the case of zero point.

When copying the data into the matrix the vector is indexed and then cast to float. Does it make sense?

Reply:
> then cast to float

No, this is wrong assumption.

```cpp
mat.create(subTargetShape.size(), subTargetShape.data(), CV_32FC1);
```

If you want to use template, you should create mat from the correct type.